### ISSUE TITLE
Version 3

### ISSUE THREAD

**Issue by southpolesteve** [Azure/azure-cosmos-js#216]
This issue will track all work related to our next major release. You can try out v3 by installing directly from github (`npm install Azure/azure-cosmos-js#v3` or the `next` tag on npm (`npm install @azure/cosmos@next`). The below list isn't in any particular order. Some of the items are blocked, are stretch goals, or may have already been completed. Anything marked as a stretch goal will not block the v3 release. No specific release date has been set.

- [x] Move to browser friendly HTTP internals using [cross-fetch](https://github.com/lquixada/cross-fetch) (Steve)
- [x] Implement request pipeline customization, bringing inline with .NET SDK (Chris)
- [x] Change response types to classes (BREAKING, Chris, Next Task)
- [x] Force setting partition key. In the near future, Cosmos will require them. (BLOCKED. Waiting on backend changes)
- [ ] Add Change Feed Processor (BLOCKED. Stretch Goal, will not block V3 release)
- [ ] Additional logging (Chris, possibly going in v2 first and ported to v3)
- [ ] Add code coverage (Stretch goal)
- [x] Enable offer tests/code (Chris, v2+v3)
- [x] Proxy via Agent (Steve, COMPLETE)
- [x] Enable no-unnecessary-classes (Steve, COMPLETE)
- [ ] Enable parallel tests  (Steve)
- [x] Split test/src into TS projects with references and enable noUnusedVariables (Steve)
- [ ] Require method return types (linting)
- [ ] Add internal flag for type defs
- [ ] Header name changes (BLOCKED by backend)
- [x] Add Cancellation tokens. Will be easy to support with fetch (#193)
- [x] 404 should not throw for reads #203 (Chris)
- [x] Remove next/current from query in favor or paginated results
- [ ] Run tests against browser again in CI (Steve)

**Comment by tony-gutierrez** [comment]
Direct mode?

**Comment by christopheranderson** [comment]
Not for the 3.0 release. 3.0 is focused on removing the partitioned experience (NP containers still work, but they will magically appear partitioned - more details soon) + some breaking change API updates to better align with other Azure SDKs & other Cosmos SDKs.

Direct Mode still doesn't have concrete plans, but I'll be discussing it during planning for the upcoming semester.

**Comment by antempus** [comment]
RE: Direct Mode: https://github.com/Azure/azure-cosmos-js/issues/47#issuecomment-518399850

**Comment by tony-gutierrez** [comment]
I can't imagine any other feature that should come before direct mode. Cosmos is expensive...and in node, its is slow...ESPECIALLY on Azure App Service. We can partition or not partition our damn collections on our own.

**Comment by southpolesteve** [comment]
Closing this issue since v3 is out and the list is fairly stale

@antempus @tony-gutierrez I'll continue any direct mode discussion over in #47
